httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java

```
@@ -60,69 +60,167 @@
+                    } else if (!gssContext.getMutualAuthState()) {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext =
+                                    HttpClientContext.adapt(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} requireMutualAuth is set but GSSAUthContext does not have"
+                                    + " mutualAuthState set", exchangeId);
+                        }
+                        state = State.FAILED;
+                        throw new AuthenticationException(
+                                "requireMutualAuth is set but GSSContext mutualAuthState is not set");
+                    } else {
+                        state = State.SUCCEEDED;
+                    }
+                }
+                break;
+            default:
+                state = State.FAILED;
+                throw new IllegalStateException("Illegal state: " + state);
+
+            }
+        } catch (final GSSException gsse) {
+            state = State.FAILED;
+            if (gsse.getMajor() == GSSException.DEFECTIVE_CREDENTIAL
+                    || gsse.getMajor() == GSSException.CREDENTIALS_EXPIRED) {
+                throw new InvalidCredentialsException(gsse.getMessage(), gsse);
+            }
+            if (gsse.getMajor() == GSSException.NO_CRED) {
+                throw new InvalidCredentialsException(gsse.getMessage(), gsse);
+            }
+            if (gsse.getMajor() == GSSException.DEFECTIVE_TOKEN
+                    || gsse.getMajor() == GSSException.DUPLICATE_TOKEN
+                    || gsse.getMajor() == GSSException.OLD_TOKEN) {
+                throw new AuthenticationException(gsse.getMessage(), gsse);
+            }
+            // other error
+            throw new AuthenticationException(gsse.getMessage(), gsse);
```

Won't those duplicate the message?

---

Not really.
But the Auth. execption would have the same message as the cause.
We can write some new message for the Auth. exception, I just really hate when a library swallows the original exception instead of wrapping it.